File: String/1786번 - 찾기/1786.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<int> get_pi(string p) {
    int len = p.length();
    vector<int> table(len, 0);
    
    int j = 0;
    for(int i = 1; i < len; i++) {
        while(j > 0 && p[i] != p[j]) j = table[j-1];
        if(p[i] == p[j]) table[i] = ++j;
    }
    
    return table;
}

vector<int> kmp(string s, string p) {
    vector<int> ans;
    vector<int> table = get_pi(p);
    int s_len = s.length();
    int p_len = p.length();
    int j = 0;
    
    for(int i = 0; i < s_len; i++) {
        while(j > 0 && s[i] != p[j]) j = table[j-1];
        if(s[i] == p[j]) {
            if(j == p_len - 1) {
                ans.push_back(i - p_len + 2);
                j = table[j];
            }
            else {
                j++;
            }
        }
    }
    
    return ans;
}

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);

	string s, p;
	getline(cin, s);
	getline(cin, p);
    
	vector<int> matched = kmp(s, p);
	cout << matched.size() << '\n';
	for(auto i : matched) cout << i << ' ';

	return 0;
}

File: String/1786번 - 찾기/1786.py
import sys

input = sys.stdin.readline

s = input().rstrip()
p = input().rstrip()

def get_pi(p):
    l = len(p)
    table = list(0 for _ in range(l + 1))
    j = 0
    
    for i in range(1, l):
        while j > 0 and p[i] != p[j]:
            j = table[j-1]
        if p[i] == p[j]:
            j += 1
            table[i] = j
            
    return table

def kmp(s, p):
    ans = []
    table = get_pi(p)
    s_len, p_len, j = len(s), len(p), 0
    for i in range(s_len):
        while j > 0 and s[i] != p[j]:
            j = table[j-1]
        if s[i] == p[j]:
            if j == p_len - 1:
                ans.append(i - p_len + 2)
                j = table[j]
            else:
                j += 1
    
    return ans
    
def solve():
    mattched = kmp(s, p)
    
    print(len(mattched))
    for i in mattched:
        print(i, end=' ')

if __name__ == '__main__':
    solve()